### ISSUE TITLE
Wrong USED disk space reported

### ISSUE THREAD

**Issue by shodanshok** [cockpit-project/cockpit-machines#1423]
As reported in a old issue for the `cockpit` project, USED disk space is wrongly reported. Original issue link: https://github.com/cockpit-project/cockpit/issues/9052

NOTE: I am opening an issue here because I think it is the more appropriate place, but feel free to close this issue if you want.

As explained in https://github.com/cockpit-project/cockpit/issues/9052#issuecomment-1918033769 the issue is about reading the wrong field from `libvirt` stats: `cockpit-machines` reads `allocation` rather than `physical`. From libvirt man page:

> block.\<num\>.allocation - offset of highest written sector in bytes
> block.\<num\>.physical - physical size of source file in bytes

So `block.<num>.physical` should be read to show USED disk space (as seen by the host).

From my understanding the fix is very simple: replace `!isNaN(vm.disksStats[target].allocation)` with `!isNaN(vm.disksStats[target].physical)` in [vmDisksCard.jsx](https://github.com/cockpit-project/cockpit-machines/blob/fe8f018c72e310a3c3c7d2752596fe78b9f488b8/src/components/vm/disks/vmDisksCard.jsx)

**Comment by mvollmer** [comment]
Hmm, for qcow2 disk images, "virsh domstats" seems to report different things with different names depending on whether the machine is running or not.

Reference is `qemu-img info` and `du`:

 - "actual size" is the size reported by `du` and by qemu-img as "disk size" in human output and "actual-size" in json. (This is the number of bytes actually used by the disk image on the host.)
 - "file length" is the size reported by `ls` and by qemu-img as "file length" in human output and "virtual-size" in json for the "file" child. (This is the size of the disk image seen by the read syscall, for example.)
 - "virtual size" is the size reported by qemu-img as "virtual size" in human output and "virtual-size" in json for the top-level. This is what was passed to `qemu create` as the size.

Cockpit should show the actual size for "Used" and the virtual size for "Capacity".  The actual size includes the qcow2 meta data overhead, I think, which is not included in the virtual size. That is acceptable, hopefully.

For a shut off machine:

 - `block.N.allocation` is the actual size.
 - `block.N.capacity` is the virtual size.
 - `block.N.physical` is the file length.

For a running machine:

 - `block.N.allocation` is a number that does not correspond to any of the qemu-img/ls/du reference numbers above. It's significantly smaller than the actual size for me, but in the [original report](https://github.com/cockpit-project/cockpit/issues/9052) it was significantly bigger. It might indeed be the "offset of highest written sector in bytes", which would be pretty useless.
 - `block.N.capacity` is the virtual size.
 - `block.N.physical` is the actual size.

Isn't it weird how `physical` changes its meaning significantly?  It looks like a bug to me, tbh.

Based on this, I'd say that for "Used", Cockpit should use `block.N.allocation` for shut off machines, but `block.N.physical` for running machines, at least for the qcow2 format.

I'll look at the raw format next, and then disk images on block devices. (I don't think we can show any useful "Used" number for block devices, right?)

**Comment by mvollmer** [comment]
And then there are external snapshots, of course, which create backing file chains and might mess up everything.

**Comment by mvollmer** [comment]
> The actual size includes the qcow2 meta data overhead, I think, which is not included in the virtual size. That is acceptable, hopefully.

Indeed: here are numbers for a disk of shut-off machine that has been fully filled inside the guest via `dd`:

```
  block.1.name=vdb
  block.1.path=/foo.qcow2
  block.1.allocation=1075724288
  block.1.capacity=1073741824
  block.1.physical=1074135040
```

Thus, the actual size is larger than the virtual size (by about 2 MB), and also larger than the file length (by about 1.5 MB).  This makes it strictly wrong to assume that the virtual size corresponds to the maximum possible value of the actual size, but it's only within a few percent, so I guess it's ok.

**Comment by mvollmer** [comment]
The "raw" format behaves like "qcow2", except:

 - For a running machine, `block.N.allocation` stays at 0.
 - There is no difference between virtual size and file length.
 - After filling it up, actual size still ends up larger than virtual size, but only by 8 kiB.

So we can use the same rules as for qcow2.

**Comment by mvollmer** [comment]
> And then there are external snapshots, of course, which create backing file chains and might mess up everything.

They do. Before creating the snapshot (shut off machine):

```
  block.0.allocation=180625408
  block.0.capacity=1073741824
  block.0.physical=1074135040
```

Machine needs 180 MB on the host to store the image for a 1 GiB disk.

After taking an external snapshot:

```
  block.0.allocation=200704
  block.0.capacity=1073741824
  block.0.physical=196624
```

Now we kinda claim the the machine needs only 200 kB to store the same content.  200 kB is only the top-most delta, however. We should show that next to the snapshot. I am not going to worry about that now.


Also with snapshots, it becomes clear that we shouldn't present "Used" and "Capacity" as related. If you have a number of snapshots, you can easily use multiple times the virtual size to store them all, and there is nothing wrong with that.

We should call those numbers "used on the host" and just "size", or something like that, to not give the impression that you run into trouble when "Used" approaches "Capacity".  "Size" is what you have specified when creating the disk, and it useful to identify them and remind you of what you have done in the past, and to spot configuration mistakes.  "Used to store it on the host" might be interesting, but honestly, I think trouble shooting the host needs a more holistic view than just looking at the details of a single VM. But showing that number is educational, at least, especially if we include snapshots in it.

**Comment by mvollmer** [comment]
I couldn't figure out how to use a block device as a disk for a VM with Cockpit Machines... Cockpit was really unhelpful, so I leave this for later.

**Comment by shodanshok** [comment]
@mvollmer thanks for looking into that issue. I agree that it is strange for `block.N.physical` to change value/meaning based on guest power status (on/off).
